sympy/ntheory/residue_ntheory.py

```
@@ -770,11 +770,18 @@ def nthroot_mod(a, n, p, all_roots=False):
     a, n, p = as_int(a), as_int(n), as_int(p)
     if n == 2:
         return sqrt_mod(a, p, all_roots)
+    # For prime moduli, handle zero residue early to include 0 as a root
+    # and avoid incorrect processing in the multiplicative group of units.
+    if not isprime(p):
+        raise NotImplementedError("Not implemented for composite p")
+
+    amod = a % p
+    if amod == 0:
+        return [0] if all_roots else 0
```

[major] The zero short-circuit now runs before `is_nthpow_residue`, so we no longer trigger the validation that used to enforce `a >= 0`, `n >= 0`, and the special-case handling for `n == 0`. On the base branch, `nthroot_mod(-17, 3, 17)` raises `ValueError('a must be >= 0')` and `nthroot_mod(0, -1, 5)` raises `ValueError('n must be >= 0')`; with this change both calls silently return `0` because `a % p == 0`. Please keep the zero return, but do it only after the existing validation has run (e.g. call `is_nthpow_residue` first, or explicitly replicate those checks) so invalid inputs still surface instead of being accepted as legitimate roots.}